### PR TITLE
264 item opinions

### PR DESCRIPTION
**Pull request by hugolopespinto**
Rework the opinion system: now opinions are linked to an item. Users can add opinions to an item, providing a note and a comment. Additionally, users can update or remove their own opinions if they created them. Users can view opinions for any item without being logged in, but they must be logged in to add an opinion.

coverage front: lines......: 63.8% (1766 of 2769 lines) 
[Screen_recording_20240317_195044.webm](https://github.com/H-chauvet/RISU/assets/72020673/4fd933e4-064d-4798-96ee-8b3e4b621d6f)
image
image
image

image
image

image
image

image

image
image

### CODE REVIEW

File: Server/controllers/Mobile/opinion.js [hunk 1]
```diff
@@ -1,37 +1,88 @@
...
+ */
+exports.updateOpinion = (opinionId, note, comment) => {
+  return db.Opinions_Mobile.update({
+    where: { id: opinionId },
+    data: {
+      note: note,
+      comment: comment
+    }
+  })
+}
```
Comment (invpuppet) [inline Server/controllers/Mobile/opinion.js:88]: missing new line

File: Server/prisma/schema.prisma [hunk 2]
```diff
@@ -111,6 +112,8 @@ model Opinions_Mobile {
     date    DateTime
```
Comment (invpuppet) [inline Server/prisma/schema.prisma:112]: maybe update Opinions_Mobile to add a real prisma relation between it and User_Mobile

**Review by invpuppet — CHANGES_REQUESTED** [review]

File: Application/mobile/app_code/lib/pages/opinion/opinion_state.dart [hunk 3]
```diff
@@ -191,10 +282,132 @@ class OpinionPageState extends State<OpinionPage> {
...
+      context: context,
+      builder: (BuildContext context) {
+        return StatefulBuilder(
+          builder: (BuildContext context, StateSetter setState) {
+            return AlertDialog(
+              title: Text(AppLocalizations.of(context)!.reviewUpdate),
+              content: Column(
+                mainAxisSize: MainAxisSize.min,
+                children: [
+                  // Champ de note
```
Comment (CedricCORGE) [inline Application/mobile/app_code/lib/pages/opinion/opinion_state.dart:299]: useless comment

File: Application/mobile/app_code/test/opinion_test.dart [hunk 4]
```diff
@@ -7,8 +7,8 @@ import 'globals.dart';
 
 void main() {
   group('Test Opinion Page', () {
-    testWidgets('find opinions pages buttons', (WidgetTester tester) async {
-      final testPage = initPage(const OpinionPage());
+    /*testWidgets('find opinions pages buttons', (WidgetTester tester) async {
```
Comment (CedricCORGE) [inline Application/mobile/app_code/test/opinion_test.dart:10]: useless commented test to remove

**Review by CedricCORGE — CHANGES_REQUESTED** [review]

**Review by invpuppet — APPROVED** [review]
good for me

**Review by CedricCORGE — APPROVED** [review]

File: Server/controllers/Mobile/user.js [hunk 5]
```diff
@@ -136,3 +136,13 @@ exports.updateUserInfo = (user, body) => {
     include: { Notifications: true }
   })
 }
+
+exports.getFullName = (userId) => {
```
Comment (31Nathan) [inline Server/controllers/Mobile/user.js:140]: Missing doc

File: Server/prisma/schema.prisma [hunk 6]
```diff
@@ -111,6 +112,9 @@ model Opinions_Mobile {
...
     note    String
     comment String   @default("")
+    itemId  Int
+    item    Item     @relation(fields: [itemId], references: [id])
```
Comment (31Nathan) [inline Server/prisma/schema.prisma:116]: Identation

File: Server/routes/Mobile/opinion.js [hunk 7]
```diff
@@ -4,10 +4,39 @@ const router = express.Router()
 const passport = require('passport')
 const userCtrl = require("../../controllers/Mobile/user")
 const opinionCtrl = require("../../controllers/Mobile/opinion")
+const itemsCtrl = require("../../controllers/Common/items")
 
-router.get('/',
+// get all opinions from an item (can be filtered by note)
```
Comment (31Nathan) [inline Server/routes/Mobile/opinion.js:9]: The doc does not go there

File: Server/routes/Mobile/opinion.js [hunk 8]
```diff
@@ -4,10 +4,39 @@ const router = express.Router()
...
+    opinions = await opinionCtrl.getOpinions(itemId, note)
+
+    return res.status(201).json({ opinions })
+  } catch (err) {
+    console.error(err.message)
+    return res.status(401).send('An error occurred')
+  }
+})
+
+// add an opinion to an item
```
Comment (31Nathan) [inline Server/routes/Mobile/opinion.js:37]: Same here

File: Server/routes/Mobile/opinion.js [hunk 9]
```diff
@@ -16,33 +45,65 @@ router.get('/',
...
-      for (var i = 0; i < opinions.length; i++) {
-        const user = await userCtrl.findUserById(opinions[i].userId)
-        result.push({
-          firstName: user.firstName ?? "Utilisateur",
-          lastName: user.lastName ?? "Supprimé",
-          comment: opinions[i].comment,
-          note: opinions[i].note
-        })
+// delete an opinion
+router.delete('/:opinionId',
```
Comment (31Nathan) [inline Server/routes/Mobile/opinion.js:74]: Again

File: Server/routes/Mobile/opinion.js [hunk 10]
```diff
@@ -16,33 +45,65 @@ router.get('/',
...
+      return res.status(201).json({ message: 'opinion deleted' })
     } catch (err) {
       console.error(err.message)
       return res.status(401).send('An error occurred')
     }
   }
 )
 
-router.post('/',
+// update an opinion
```
Comment (31Nathan) [inline Server/routes/Mobile/opinion.js:105]: And Again

**Review by 31Nathan — CHANGES_REQUESTED** [review]

**Review by 31Nathan — APPROVED** [review]

**Pull request merged**